### DISCUSSION TITLE
Hi, does PropertyGrid support images?

### DISCUSSION THREAD

**Discussion by shenshengyi** [iTwin/itwinjs-core#945] [Q&A] [answered]
Hi, does PropertyGrid support embedding images?

Or can I only pop up the control that displays the picture by triggering the event?  Thanks.

**Comment by grigasp** [comment, marked as answer]
Hi,

First, I'd like to mention that we have two property grid components: 
- [PropertyGrid](https://github.com/imodeljs/imodeljs/blob/b010d2913a3167dc47680c384258c2c700743d7c/ui/components/src/ui-components/propertygrid/component/PropertyGrid.tsx#L66)
- [VirtualizedPropertyGrid](https://github.com/imodeljs/imodeljs/blob/b010d2913a3167dc47680c384258c2c700743d7c/ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGrid.tsx#L127)

We're not actively working on the first one and are planning to deprecate it, so you should be looking at the virtualized property grid, which renders only visible properties and can performantly handle much larger number of them.

There're two basic concepts that property grid renders - [PropertyCategory](https://github.com/imodeljs/imodeljs/blob/b010d2913a3167dc47680c384258c2c700743d7c/ui/components/src/ui-components/propertygrid/PropertyDataProvider.ts#L16) and [PropertyRecord](https://github.com/imodeljs/imodeljs/blob/b010d2913a3167dc47680c384258c2c700743d7c/ui/abstract/src/ui-abstract/properties/Record.ts#L33). At this moment customizations are only allowed for [PropertyRecord](https://github.com/imodeljs/imodeljs/blob/b010d2913a3167dc47680c384258c2c700743d7c/ui/abstract/src/ui-abstract/properties/Record.ts#L33) rendering - that can be done with these steps:
- Assign `record.property.renderer.name` to a name of your custom renderer.
- Register your custom renderer using [PropertyValueRendererManager.defaultManager.registerRenderer](https://github.com/imodeljs/imodeljs/blob/b010d2913a3167dc47680c384258c2c700743d7c/ui/components/src/ui-components/properties/ValueRendererManager.tsx#L132) function.

This will make the value of the property to be rendered using your custom renderer. You're free to decide how it renders the value - it can be either a link that opens a dialog, an image, or anything else.

**Reply by shenshengyi** [reply]
Thank you very much. Your answers have helped me a lot.
